src/coord/src/coord.rs

```diff
@@ -479,6 +479,25 @@ where
                         self.new_frontiers(entry.id(), Some(0), self.logical_compaction_window_ms);
                     self.sources.insert(entry.id(), frontiers);
                 }
+                CatalogItem::Table(table) => {
+                    let since_ts = {
+                        match &table.persist {
+                            Some(persist) => Some(persist.since_ts),
+                            _ => None,
+                        }
+                    };
+
+                    let since_ts = since_ts.unwrap_or(0);
+                    let frontiers = self.new_frontiers(
+                        entry.id(),
+                        [since_ts],
+                        self.logical_compaction_window_ms,
+                    );
+
+                    // NOTE: Tables are not sources, but to a large part of the system they look
+                    // like they are, e.g. they are rendered as a SourceConnector::Local.
+                    self.sources.insert(entry.id(), frontiers);
```

Comment:
Does this mean that `update_upper` has been getting messages for both a table and its index, but has been ignoring the non-index message? Or does the since setting happen only once here? Can you explain a bit more (or add a test) that illustrates the problem this is solving?

Reply:
Ok, that's useful context. I agree with your recommendation.

Reply:
@mjibson There are existing tests for this, basically all tests under `test/persistence/user-tables` (see https://github.com/MaterializeInc/materialize/tree/main/test/persistence/user-tables). The idea of the tests is that they do a thing, then `materialized` is stopped and restarted and then the `after` tests are run to verify.

Persistent user tables are like sources in a way, a source that reads from persistence is created to read from the underlying persistent stream. What I did, is change the API to force `dataflow` to hand the expected `as_of_frontier` to the `replay()` operator (this is the thing that replays persisted data), and to verify that the `since` (aka compaction frontier) of the persisted stream is compatible with the requested `as_of_frontier`. This failed all the tests, because we're not correctly calculating an `as_of_frontier`, because we're not taking the `since` of the persisted stream into account. The first commit in this PR fixes that, to make the tests pass again.

@ruchirK It is _a_ valid place to put it: with that line, we correctly determine a since and tests pass, without it they don't.

Here is where the since that we put in when bootstrapping is picked up: https://github.com/MaterializeInc/materialize/blob/b4e67269dc56279ede1f6a4c7e52dddc3c51fdce/src/coord/src/coord.rs#L4118 And this is where the frontier for the index that is rendered for the table is registered: https://github.com/MaterializeInc/materialize/blob/b4e67269dc56279ede1f6a4c7e52dddc3c51fdce/src/coord/src/coord.rs#L4143

The workflow (when restarting is roughly this):

1. coordinator bootstraps its metadata, **which now includes reading the persistence since for tables and registering it with the source frontiers**
2. coordinator notices that it needs to render a dataflow for the table
3. the code I linked to above is consulted to determine a since
4. the index frontier of the arrangement that is about to be rendered is registered
5. the "source" is being rendered.

So, as I mentioned above, to a lot of the system the table looks like a source. Not sure that's good, though... 😅

Reply:
That makes sense. This construction is a little bit nonintuitive to me though because we indirectly use the table's `since` on startup to initialize the primary index's `since` and then never update it again. That happens to work implicitly because we know once we've rendered the primary index, we'll never use the table's `since` again, but it feels very action-at a distance.

It seems like it is more work to directly initialize the index's since frontier, so I'm fine with this, but perhaps would you consider also updating the table source's since frontier when we allow compaction on the index/persisted data and setting the table source's since frontier to zero when we first create it? The way the code is structured now, we only keep an entry in `self.sources` for a table id after a restart which i suspect is correct right now, but might lead to surprising effects in the future

Reply:
I didn't do that initially because of the weird nature of Table "sources", but I will do that! I think it might even fix a bug with disabling/enabling user-indexes. I will have to pick that up again in the new year, though.